openmls/src/binary_tree/array_representation/diff.rs

```diff
@@ -0,0 +1,634 @@
+
+        // We want to return the position of the lowest common ancestor in the
+        // direct path of `leaf_index_1` (i.e. the sender_leaf_index).
+        let subtree_root_node_index =
+            lowest_common_ancestor(to_node_index(leaf_index_1), to_node_index(leaf_index_2));
+
+        // Figure out which child is the relevant copath node.
+        let copath_node_index = if leaf_index_2 < leaf_index_1 {
+            left(subtree_root_node_index)?
+        } else {
+            right(subtree_root_node_index, self.size())?
+        };
+
+        NodeReference::try_from_node_index(self, copath_node_index)
+    }
+
+    /// Returns a vector of [`NodeReference`]s, where the first reference is to
+    /// the root of the shared subtree of the two given leaf indices followed by
+    /// references to the nodes in the direct path of said subtree root.
+    ///
+    /// Returns an error if either of the two given leaf indices do not
+    /// correspond to a leaf in the diff.
+    pub(crate) fn subtree_path(
+        &self,
+        leaf_index_1: LeafIndex,
+        leaf_index_2: LeafIndex,
+    ) -> Result<Vec<NodeReference>, ABinaryTreeDiffError> {
+        let node_index_1 = to_node_index(leaf_index_1);
+        let node_index_2 = to_node_index(leaf_index_2);
+
+        self.out_of_bounds(node_index_1)?;
+        self.out_of_bounds(node_index_2)?;
+
+        let lca = lowest_common_ancestor(node_index_1, node_index_2);
+        let direct_path_indices = direct_path(lca, self.size())?;
+        let mut full_path = vec![NodeReference::try_from_node_index(self, lca)?];
+        for node_index in direct_path_indices {
+            let node_ref = NodeReference::try_from_node_index(self, node_index)?;
+            full_path.push(node_ref);
+        }
+
+        Ok(full_path)
+    }
+
+    // Functions pertaining to the whole diff
+    /////////////////////////////////////////
+
+    /// Returns an iterator over references to the content of all nodes in the
+    /// diff.
+    pub(crate) fn iter(&'a self) -> DiffIterator<'a, T> {
+        DiffIterator {
+            diff: self,
+            current_index: 0u32,
+        }
+    }
+
+    /// Returns a vector containing the nodes of the tree in-order, i.e. in the
+    /// array representation of the diff. This function should not fail and only
+    /// returns a [`Result`], because it might throw a
+    /// [LibraryError](ABinaryTreeError::LibraryError).
+    pub(crate) fn export_nodes(&self) -> Result<Vec<T>, ABinaryTreeDiffError> {
+        let mut nodes = Vec::new();
+        for node_index in 0..self.size() {
+            let node = self.node_by_index(node_index)?;
+            nodes.push(node.clone());
+        }
+        Ok(nodes)
+    }
+
+    /// Returns the size of the diff.
+    pub(in crate::binary_tree) fn size(&self) -> NodeIndex {
+        self.size
+    }
+
+    /// Returns the leaf count of the diff.
+    pub(crate) fn leaf_count(&self) -> LeafIndex {
+        (self.size() + 1) / 2
```

Comment:
We need to check at least with a `debug_assert` that this doesn't overflow (in case you're certain that it shouldn't).